### DISCUSSION TITLE
Decide testing framework

### DISCUSSION THREAD

**Discussion by fardavide** [androiddevelopersitalia/aditogether#20] [General]
# Problem
We discussed that JUnit works only under JVM.
We need a testing framework that works for Kotlin Multiplatform and supports all the targets needed by the project.

# Possible solutions
* [Kotest](https://kotest.io/)

**Comment by fardavide** [comment]
We're experimenting with Kotest here https://github.com/androiddevelopersitalia/aditogether/compare/main...feat/21_detekt-rules-setup

As of now, it seems the most viable solution, as widely used and constantly maintained and updated.
Additionally it runs on Junit5, which lets us keep a door open for Mutation Testing in case we want to consider that in the future (https://gradle-pitest-plugin.solidsoft.info/)

I think we can give it a try and see how it plays

CC @mcatta @fondesa

**Comment by mcatta** [comment]
Kotest looks cool, I think it could fit very well for the entire shared modules, and I see that it supports also [Arrow's monads](https://kotest.io/docs/assertions/arrow.html).

The syntax seems a bit different compared with the classic Junit style, as far as I understand, we don't have a centralized `BeforeTest/AfterTest` but it should be defined inside the `Given` condition. This means that if we need to write something similar to a `org.junit.Rule` we need to create something on of that, isn't it?

**Reply by fardavide** [reply]
We also have `AnnotationSpec` which reflects the classic style: https://kotest.io/docs/framework/testing-styles.html#annotation-spec

But honestly I'm curious to try another style (particularly the behavioral)

**Comment by fondesa** [comment]
We can try Kotest in my opinion.
The only thing I'd be strict about is using the same [testing style](https://kotest.io/docs/framework/testing-styles.html) in our repo for consistency reasons.

**Reply by fardavide** [reply]
100% agree, I hope it's possible to import the dependency for a single style, I didn't check yet.

Besides that, at worst we would need the Annotation Style for some cases, but we need to verify that, I'm not familiar with the paradigm

**Reply by fondesa** [reply]
> 100% agree, I hope it's possible to import the dependency for a single style, I didn't check yet.

In case if it is not, we can create a lint rule to forbid the usages of the other styles

**Reply by fardavide** [reply]
Actually, there is already a `ForbiddenImport` in Detekt

**Reply by fondesa** [reply]
Even better!
